msgp/read_bytes.go

```diff
@@ -397,6 +397,158 @@ func ReadDurationBytes(b []byte) (d time.Duration, o []byte, err error) {
 	return time.Duration(i), o, err
 }
 
+// ReadInt64BytesCanonical tries to read an int64
+// from 'b' and return the value and the remaining bytes.
+// The encoding must be canonical and an error is returned if it isn't.
+// Possible errors:
+// - ErrShortBytes (too few bytes)
+// - ErrNonCanonical (not canonical encoding)
+// - TypeError (not a int)
+func ReadInt64BytesCanonical(b []byte) (i int64, o []byte, err error) {
+	l := len(b)
+	if l < 1 {
+		return 0, nil, ErrShortBytes
+	}
+
+	lead := b[0]
+	if isfixint(lead) {
+		i = int64(rfixint(lead))
+		o = b[1:]
+		return
+	}
+	if isnfixint(lead) {
+		i = int64(rnfixint(lead))
+		o = b[1:]
+		return
+	}
+
+	switch lead {
+	case mnil:
+		i = 0
+		o = b[1:]
+		return
+
+	case mint8:
+		if l < 2 {
+			err = ErrShortBytes
+			return
+		}
+		i = int64(getMint8(b))
+		o = b[2:]
+		if (i << 7) == 0 {
```

Comment:
I think this should be `i << 5`, because msgpack supports only 5-bit negative fixint.

Reply:
Actually, I think there might be two other issues here.  First, it seems like you should be shifting right (i.e., `i >> 5`).  Second, for signed values, I think you should check for `== -1` rather than `== 0`.

Reply:
Thanks for the catch, hopefully would have caught it soon in testing. JW how much of a performance vs readability tradeoff would it be to compare against int64 constants in this case rather than doing the bitshift comparisons?

Reply:
I doubt there's a measurable performance difference.

Just out of curiosity, I tested go 1.21.0, and the compiler emits two instructions (`CMPQ` and `SETCS`) for `x < 256`, and emits three instructions (`SHRQ`, `TESTQ`, `SETEQ`) for `(x >> 8) == 0`.  I didn't bother benchmarking them; we should go for what's clearer.

I don't have a strong opinion about whether bit-shifts or comparisons are clearer here.